homeassistant/components/schlage/config_flow.py

```diff
@@ -16,43 +18,97 @@
-                errors["base"] = "invalid_auth"
-            except Exception:  # pylint: disable=broad-except
-                LOGGER.exception("Unknown error")
-                errors["base"] = "unknown"
-            else:
-                await self.async_set_unique_id(user_id)
-                return self.async_create_entry(title=username, data=user_input)
+        if user_input is None:
+            return self._show_user_form(errors)
+        username = user_input[CONF_USERNAME]
+        password = user_input[CONF_PASSWORD]
+        user_id = await self.hass.async_add_executor_job(
+            _authenticate, username, password, errors
+        )
+        if errors:
+            return self._show_user_form(errors)
+
+        await self.async_set_unique_id(user_id)
+        return self.async_create_entry(title=username, data=user_input)
 
+    def _show_user_form(self, errors: dict[str, str]) -> FlowResult:
+        """Show the user form."""
         return self.async_show_form(
             step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
         )
 
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Handle reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
 
-def _authenticate(username: str, password: str) -> str:
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that reauth is required."""
+        assert self.reauth_entry is not None
+        errors: dict[str, str] = {}
+        if user_input is None:
+            return self._show_reauth_form(errors)
+
+        username = self.reauth_entry.data[CONF_USERNAME]
+        password = user_input[CONF_PASSWORD]
+        user_id = await self.hass.async_add_executor_job(
+            _authenticate, username, password, errors
+        )
+        if errors:
+            return self._show_reauth_form(errors)
+
+        if self.reauth_entry.unique_id != user_id:
+            return self.async_abort(reason="wrong_account")
+
+        data = {
+            CONF_USERNAME: username,
+            CONF_PASSWORD: user_input[CONF_PASSWORD],
+        }
+        self.hass.config_entries.async_update_entry(self.reauth_entry, data=data)
+        await self.hass.config_entries.async_reload(self.reauth_entry.entry_id)
+        return self.async_abort(reason="reauth_successful")
+
+    def _show_reauth_form(self, errors: dict[str, str]) -> FlowResult:
+        """Show the reauth form."""
+        return self.async_show_form(
+            step_id="reauth_confirm",
+            data_schema=STEP_REAUTH_DATA_SCHEMA,
+            errors=errors,
+        )
+
+
+def _authenticate(username: str, password: str, errors: dict[str, str]) -> str | None:
```

Comment:
Should this not return the errors if any?

Reply:
It modifies the `errors` dict that's passed as an argument.

Reply:
Right, don't like but I guess that's fine.

Reply:
I changed it to return the errors instead of taking it as a parameter.